enhancements/20250522-expose-cdi-settings-in-ui.md

```diff
@@ -0,0 +1,114 @@
+
+### User Experience In Detail
+
+The CDI settings will be integrated into the existing Storage Classes page in Harvester UI.
+
+### API changes
+
+## Design
+
+### Implementation Overview
+
+#### UI Preview
+
+![storageclass-ui](./20250522-expose-cdi-settings-in-ui/storageclass-ui.png)
+
+- Note: These new tab: CDI Settings is not available in the longhorn-v1 StorageClass.
+- The new CDI settings can be configured during the creation of a StorageClass or by updating a StorageClass.
+- The Volume Mode / Access Modes setting supports multiple combinations, where each combination must have a unique Volume Mode and one or more associated Access Modes.
+  - The volume mode should be either `FileSystem` or `Block`.
+  - The access modes can have 0~4 unique mode, the access mode could be either `ReadWriteOnce`, `ReadOnlyMany`, `ReadWriteMany` or `ReadWriteOncePod`.
+- The Volume Snapshot Class is presented as a dropdown menu listing all available VolumeSnapshotClass names.
+- The Clone Strategy is also a dropdown menu with the following fixed options: "copy", "snapshot", and "csi-clone".
+- The File System Overhead must be a value between 0 and 1, with a regex pattern of ^(0(?:\.\d{1,3})?|1)$ (i.e., up to three digits after the decimal point).
+
+#### Backend Implementation
+
+Introduce 4 annotations to StorageClass:
+
+- `cdi.harvesterhci.io/filesystemOverhead`: String value representing filesystem overhead percentage (0-1), regex pattern: `^(0(?:\.\d{1,3})?|1)$` (up to 3 digits after the decimal point)
+- `cdi.harvesterhci.io/cloneStrategy`: String value, one of: "copy", "snapshot", "csi-clone"
+- `cdi.harvesterhci.io/volumeSnapshotClass`: String value, must be an existing VolumeSnapshotClass name
+- `cdi.harvesterhci.io/volumeModeAccessModes`: JSON string mapping VolumeMode to AccessMode arrays
+
+Below is the StorageClass yaml example:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+    cdi.harvesterhci.io/filesystemOverhead: "0.85"
+    cdi.harvesterhci.io/cloneStrategy: "snapshot"
+    cdi.harvesterhci.io/volumeSnapshotClass: "csi-snapshot-class"
+    cdi.harvesterhci.io/volumeModeAccessModes: |
+      {"Block":["ReadWriteOnce","ReadOnlyMany"],"Filesystem":["ReadWriteOnce","ReadWriteMany"]}
+```
+
+The implementation consists of three main components:
+
+1. StorageClass Webhook Validation
+Extend the existing StorageClass validation webhook to validate the five new annotations. The webhook will perform pattern matching for filesystemOverhead, enum validation for cloneStrategy, existence checks for volumeSnapshotClass, and JSON schema validation for volumeModeAccessModes.
+2. StorageProfile Controller
+Modify the existing storageprofile_controller to watch for StorageClass changes and synchronize cloneStrategy, volumeSnapshotClass, and volumeModeAccessModes annotations to the corresponding StorageProfile CRD.
```

Comment:
Check if it's feasible to only monitor the annotated storage classes to avoid watching/reconciling all storage classes.

Reply:
As far as I know, Wrangler doesn't support this scenario. What we can do is return early in the controller if the given object doesn't contain the specific annotation.
for example:
```go
ctrl.OnChange(ctx, "sc-handler", func(key string, obj *storagev1.StorageClass) (*storagev1.StorageClass, error) {
    if obj == nil {
        return nil, nil
    }
    if _, ok := obj.Annotations["my-annotation"]; !ok {
        return obj, nil
    }
    ...
```